### PR TITLE
ZED SDK 3.0 Refactor 

### PR DESCRIPTION
**Pull request by mmartin4972**
closes Issue #334 
Refactor for ZED SDK 3.0
Recording of Obstacle Detection
Recording of AR Tag Detection

### CODE REVIEW

File: onboard/cv/artag_detector.cpp [hunk 1]
```diff
@@ -34,6 +34,8 @@ TagDetector::TagDetector() {  //initializes detector object with pre-generated d
     // initialize other special parameters that we need to properly detect the URC (Alvar) tags
     alvarParams = new cv::aruco::DetectorParameters();
     alvarParams->markerBorderBits = 2;
+    alvarParams->doCornerRefinement = false; //change 1
```
Comment (calebhar12) [inline onboard/cv/artag_detector.cpp:37]: **Nit** What does `change 1` mean here? If it's a setting that may need to be changed by others during testing, make sure the comment is more clear.

File: onboard/cv/camera.cpp [hunk 2]
```diff
@@ -28,44 +37,52 @@ class Camera::Impl {
...
-	assert(this->zed_.open(init_params) == sl::SUCCESS);
-	this->zed_.setConfidenceThreshold(THRESHOLD_CONFIDENCE);
+
+	assert(this->zed_.open() == sl::ERROR_CODE::SUCCESS);
+  
+  //Parameters for Positional Tracking
+  init_params.coordinate_system = sl::COORDINATE_SYSTEM::RIGHT_HANDED_Y_UP; // Use a right-handed Y-up coordinate system
+  
+  //weird zed shit
+  // this->zed_.setCameraSettings(sl::CAMERA_SETTINGS_EXPOSURE, 100, true);
```
Comment (calebhar12) [inline onboard/cv/camera.cpp:52]: **Nit:** remove this lol

File: onboard/cv/config.h.in [hunk 3]
```diff
@@ -1,7 +1,9 @@
 #pragma once
-#mesondefine TB_DETECTION
+#mesondefine AR_DETECTION
```
Comment (calebhar12) [inline onboard/cv/config.h.in:2]: **Nit:** group these more nicely, i.e. have `AR_DETECTION` followed by `AR_RECORD` followed by `OBSTACLE_DETECTION` followed by `OBS_RECORD`. Just so it's clear in this file what actually matters, in case we have to mess w these late on.

File: onboard/cv/main.cpp [hunk 4]
```diff
@@ -125,13 +188,14 @@ int main() {
           arTags[0].id = -1;
         }
       } else { //one tag found
-        arTags[0].distance = depth_img.at<float>(tagPair.first.loc.y, tagPair.first.loc.x);
+        if(!isnan(depth_img.at<float>(tagPair.first.loc.y, tagPair.first.loc.x)))
```
Comment (calebhar12) [inline onboard/cv/main.cpp:191]: What happens if this value is `nan`? do we then never set a bearing/distance/id for tag 0? 

It's my understanding that even if at the center of the tag our depth happens to be `nan`, the depth could till be `non-nan` for other points in the tag. I think in this case where `depth_img` at the centerpoint of the tag is `nan`, we should probably search for some other points on the tag that may not be nan, instead of just scrapping tag 0's information altogether.

File: onboard/cv/main.cpp [hunk 5]
```diff
@@ -125,13 +188,14 @@ int main() {
...
+          arTags[0].distance = depth_img.at<float>(tagPair.first.loc.y, tagPair.first.loc.x);
```
Comment (calebhar12) [inline onboard/cv/main.cpp:192]: **Nit:** Even though there's only one line of code in the if-statement, I'd add brackets for readability. 

Also, in the future we should consider adding some method for searching around the tag's centerpoint for a non-`nan` depth value-- It wouldn't be too hard to do I think

File: onboard/cv/meson_options.txt [hunk 6]
```diff
@@ -1,6 +1,9 @@
-option('tb_detection', type: 'boolean', value : true)
+option('ar_detection', type: 'boolean', value : true)
```
Comment (calebhar12) [inline onboard/cv/meson_options.txt:1]: **Nit:** As I mentioned above, change the grouping here

File: onboard/cv/obstacle_detector.cpp [hunk 7]
```diff
@@ -86,7 +107,8 @@ pair<int, float> get_final_col(vector<pair<int, float> > & sorted_sums, float mi
     // go straight if each direction has similar value
     return make_pair(-1, middle_sum);
 
-  } else {
+  } 
+
```
Comment (calebhar12) [inline onboard/cv/obstacle_detector.cpp:111]: Why did you remove the `else` here? (Just curious, I have no issue with it at the moment but just wondering b/c this seems like it might be a logic change, as opposed to a refactoring change)

File: onboard/cv/artag_detector.cpp (hunk 1 above)
Comment (mmartin4972) [reply onboard/cv/artag_detector.cpp:37]: Not sure. Probably an old comment

File: onboard/cv/camera.cpp (hunk 2 above)
Comment (mmartin4972) [reply onboard/cv/camera.cpp:52]: For sure

File: onboard/cv/config.h.in (hunk 3 above)
Comment (mmartin4972) [reply onboard/cv/config.h.in:2]: Alrighty

File: onboard/cv/main.cpp (hunk 5 above)
Comment (mmartin4972) [reply onboard/cv/main.cpp:192]: Absofruitly. Sounds like a good plan.

File: onboard/cv/meson_options.txt (hunk 6 above)
Comment (mmartin4972) [reply onboard/cv/meson_options.txt:1]: Np. I see you're feeling nitty today.

File: onboard/cv/obstacle_detector.cpp (hunk 7 above)
Comment (mmartin4972) [reply onboard/cv/obstacle_detector.cpp:111]: I don't think it changed the logic, since the else was empty. I thought it just seemed cleaner without it.

File: onboard/cv/camera.cpp [hunk 8]
```diff
@@ -2,9 +2,18 @@
 #include "perception.hpp"
 
 #if ZED_SDK_PRESENT
+
+#pragma GCC diagnostic ignored "-Wreorder" //Turns off warning checking for sl lib files
```
Comment (raytitan) [inline onboard/cv/camera.cpp:6]: There's no way around this? Seems like Travis might be ignoring it

File: onboard/cv/main.cpp [hunk 9]
```diff
@@ -70,15 +71,65 @@ void write_curr_frame_to_disk(Mat rgb, Mat depth, int counter) {
...
   #if PERCEPTION_DEBUG
-    namedWindow("image", 1);
+    namedWindow("Obstacle");
     namedWindow("depth", 2);
   #endif
   disk_record_init();
 
+  //Video Stuff
+  TagDetector d1;
+  pair<Tag, Tag> tp;
```
Comment (raytitan) [inline onboard/cv/main.cpp:85]: NIT: tp = tag pair? This section could have more descriptive variable names, but just a nit

File: onboard/cv/camera.cpp (hunk 8 above)
Comment (calebhar12) [reply onboard/cv/camera.cpp:6]: also-- is it failing on warning checking? Or are we just disabling it because warning checking is messy/looks bad?

File: onboard/cv/obstacle_detector.cpp (hunk 7 above)
Comment (calebhar12) [reply onboard/cv/obstacle_detector.cpp:111]: I don't think the else was empty? I looked here: https://github.com/umrover/mrover-workspace/blob/master/onboard/cv/obstacle_detector.cpp#L89 and it seems like lines 89-101 were all contained within the else statement.

File: onboard/cv/camera.cpp [hunk 10]
```diff
@@ -28,44 +37,52 @@ class Camera::Impl {
...
+		                   sl::MAT_TYPE::F32_C1);
 	this->depth_ = cv::Mat(
 		this->image_size_.height, this->image_size_.width, CV_32FC1,
-		this->depth_zed_.getPtr<sl::uchar1>(sl::MEM_CPU));
+		this->depth_zed_.getPtr<sl::uchar1>(sl::MEM::CPU));
 }
 
 bool Camera::Impl::grab() {
-	return this->zed_.grab(this->runtime_params_) == sl::SUCCESS;
+  return this->zed_.grab() == sl::ERROR_CODE::SUCCESS;
```
Comment (DavidSmith166) [inline onboard/cv/camera.cpp:74]: **Nit:** is there a reason why this return statement doesn't follow the same indentation as the rest of the file?

File: onboard/cv/main.cpp [hunk 11]
```diff
@@ -70,15 +71,65 @@ void write_curr_frame_to_disk(Mat rgb, Mat depth, int counter) {
...
+  tp = d1.findARTags(src, depth_img, rgb);
+  Size fsize = rgb.size();
+  
+  string s = "artag_" + timeStamp + ".avi";
+
+  VideoWriter vidWrite(s, VideoWriter::fourcc('M','J','P','G'),10,fsize,true);
+
+  if(vidWrite.isOpened() == false)
+  {
+	  cout << "didn't open";
```
Comment (DavidSmith166) [inline onboard/cv/main.cpp:108]: **NIT:** would it not be more appropriate to write this to stderr and add a newline?

File: onboard/cv/main.cpp [hunk 12]
```diff
@@ -70,15 +71,65 @@ void write_curr_frame_to_disk(Mat rgb, Mat depth, int counter) {
...
+
+  avoid_obstacle_sliding_window(bigD, src, num_sliding_windows, roverpw);
+  Size fs = src.size();
+  string name = "obs_" + timeStamp + ".avi";
+
+  VideoWriter vidWriteObs(name, VideoWriter::fourcc('M','J','P','G'),10,fs,true);
+
+  if(vidWriteObs.isOpened() == false)
+  {
+	  cout << "didn't open";
```
Comment (DavidSmith166) [inline onboard/cv/main.cpp:128]: **NIT:** see my other comment on error handling

File: onboard/cv/main.cpp [hunk 13]
```diff
@@ -156,35 +221,40 @@ int main() {
...
     #endif
 
     j++;
   }
-
+  #if AR_RECORD
+  vidWrite.release();
+  #endif
+  #if OBS_RECORD
+  vidWriteObs.release();
```
Comment (DavidSmith166) [inline onboard/cv/main.cpp:257]: **NIT:** You may want to consider a wrapper around the VideoWriter objects that releases them upon destruction. This would remove the need for this block of release statements and RAII is generally a good practice to follow. That said I'm not super familiar with the Perception codebase so lmk if this wouldn't make sense

File: onboard/cv/obstacle_detector.cpp [hunk 14]
```diff
@@ -2,10 +2,20 @@
...
 using namespace std;
 
+//Global Variables 
+int THRESHOLD_NO_OBSTACLE_CENTER = 300000;
+int THRESHOLD_NO_SUBWINDOW = 70000;
 // record the last direction
 static int last_center;
 Rect cropped = Rect( 20, SKY_START_ROW, 1240, 300 ); // (x, y, width, height) 
 
+void updateThresholds(int in1, int in2)
```
Comment (DavidSmith166) [inline onboard/cv/obstacle_detector.cpp:12]: **NIT:** You can use more descriptive parameter names here

File: onboard/cv/main.cpp (hunk 13 above)
Comment (calebhar12) [reply onboard/cv/main.cpp:257]: ^+1 to this. Having a wrapper to release the objects makes sense, and potentially we could also abstract away some of the logic related to the VideoWriter objects, which I think could make this file cleaner overall. Up to you though Matt, this could also be part of a later refactor if you think it doesn't make sense to change this now.

File: onboard/cv/camera.cpp (hunk 8 above)
Comment (mmartin4972) [reply onboard/cv/camera.cpp]: We're just doing it because it looks bad. There are a few camera resolution warnings that are thrown that I have been having a hard time getting rid of that you're probably seeing on Travis.

File: onboard/cv/main.cpp (hunk 13 above)
Comment (mmartin4972) [reply onboard/cv/main.cpp]: I'll add it as a new member task for the future

**Review by raytitan — APPROVED** [review]

**Review by DavidSmith166 — APPROVED** [review]
